src/mscorlib/src/System/IO/StreamReader.cs

```diff
@@ -237,7 +244,23 @@ internal StreamReader(String path, Encoding encoding, bool detectEncodingFromByt
             byteLen = 0;
             bytePos = 0;
             _detectEncoding = detectEncodingFromByteOrderMarks;
-            _preamble = encoding.GetPreamble();
+
+            // Encoding.GetPreamble() always allocates and returns a new byte[] array for
+            // encodings that have a preamble.
+            // We can avoid repeated allocations for the default and commonly used Encoding.UTF8
+            // encoding by using our own private cached instance of the UTF8 preamble.
+            // We specifically look for Encoding.UTF8 because we know it has a preamble,
+            // whereas other instances of UTF8Encoding may not have a preamble enabled, and
+            // there's no public way to tell if the preamble is enabled for an instance other
+            // than calling GetPreamble(), which we're trying to avoid.
+            // This means that other instances of UTF8Encoding are excluded from this optimization.
+            // We check if the encoding is UTF8Encoding before the reference equality comparison
+            // with Encoding.UTF8, to avoid the potential first-time Encoding.UTF8 allocation
+            // when the encoding isn't even UTF8Encoding.
+            _preamble = (encoding is UTF8Encoding && object.ReferenceEquals(encoding, Encoding.UTF8)) ?
```

Comment:
I tend to agree with Wes on this one... the chances that Encoding.UTF8 is not initialized during the program's execution is very unlikely. I've not tested on CoreCLR, but with the full framework it appears that Encoding.utf8Encoding is already allocated by the start of Main.

(Separately, as Encoding.UTF8 is lazily initialized without synchronization, it's possible that a race condition could happen where the encoding passed in was retrieved from Encoding.UTF8 but is different than the object returned from this call site to Encoding.UTF8, but the chances of that happening are low and the only negative effect is that we get the allocation we currently have and are trying to avoid with this change).